src/osx/cocoa/dataview.mm

```diff
@@ -819,35 +820,38 @@ -(NSDragOperation) setupAndCallDataViewEvents:(wxEventType)eventType dropInfo:(i
             // clean-up:
             delete dataObjects;
         }
+
+        return dragOperation;
+    }
+
+    wxDataObjectComposite* dataObjects(new wxDataObjectComposite());
+    wxDataObjectSimple* dataObject;
+
+    dragOperation = wxDragError;
+    if ([bestType compare:NSFilenamesPboardType] == NSOrderedSame)
+    {
+        dataObject = new wxFileDataObject;
+    }
+    else if ([bestType compare:NSStringPboardType] == NSOrderedSame)
+    {
+        dataObject = new wxTextDataObject;
     }
     else
     {
-        // needed to convert internally used UTF-16 representation to a UTF-8
-        // representation
-        CFDataRef              osxData;
-        wxDataObjectComposite* dataObjects   (new wxDataObjectComposite());
-        wxTextDataObject*      textDataObject(new wxTextDataObject());
-
-        osxData = ::CFStringCreateExternalRepresentation(kCFAllocatorDefault,(CFStringRef)[pasteboard stringForType:NSStringPboardType],
-#if defined(wxNEEDS_UTF16_FOR_TEXT_DATAOBJ)
-                                                         kCFStringEncodingUTF16,
-#else
-                                                         kCFStringEncodingUTF8,
-#endif
-                                                         32);
-        if (textDataObject->SetData(::CFDataGetLength(osxData),::CFDataGetBytePtr(osxData)))
-            dataObjects->Add(textDataObject);
-        else
-            delete textDataObject;
-        // send event if data could be copied:
+        dragOperation = wxDragNone;
+    }
 
-        dragOperation = [self callDataViewEvents:eventType dataObjects:dataObjects item:item proposedChildIndex:index];
+    if (dataObject)
+    {
+        dataObjects->Add(dataObject);
 
-        // clean up:
-        ::CFRelease(osxData);
-        delete dataObjects;
+        wxOSXPasteboard wxPastboard(pasteboard);
+        if (dataObject->ReadFromSource(&wxPastboard))
```

Comment:
This approach seems to be invalid in case of dragging text. 😔 I received only the first letter of dragged Latin string or invalid symbols of dragged Cyrillic string.

Reply:
Looks like a wrong cast from `wchar_t*` to `char*` somewhere.

Reply:
@csomor can you help me with this issue, please? Can I use `ReadFromSource` in case of dragging text here?

Reply:
I'll try to look at it over the weekend

Reply:
usually when using 'internal sources' I don't go via the Pasteboard at all, I call the `wxDropSource::GetCurrentDropSource` and then read from that object (see `dnd_osx.cpp`, `wxDropTarget::GetData()`), I'll dive deeper

(btw. please keep the structure of having an else branch and no early return in the if branch)

Reply:
ok I see that dataview.mm is setting up the strings itself (I didn't write dataview.mm), what are the steps I need to perform to view the problem exactly ? Thanks

Reply:
I found the bug, the code is assuming we are always using UTF8 internally, in line 927 in dataview.mm:

```
 itemString = wxString::FromUTF8(static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId),dataSize)
```

Reply:
I'd suggest replacing this line with something along, @vadz is this fine for you ?
```
static wxMBConvUTF16 s_UTF16Converter;
itemString = wxString( static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId), s_UTF16Converter, dataSize );
```

Reply:
> ok I see that dataview.mm is setting up the strings itself (I didn't write dataview.mm), what are the steps I need to perform to see the problem exactly ? Thanks

You can run the dataview sample and do dnd inside the app and drop text from outside.

Reply:
Thanks, I found out I could drag Beethoven around :-)

Reply:
> 
> 
> I'd suggest replacing this line with something along this, @vadz is this fine for you ?
> 
> ```
> static wxMBConvUTF16 s_UTF16Converter;
> itemString = wxString( static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId), s_UTF16Converter, dataSize );
> ```

Sorry, I didn't really have time to look at this code yet, but why do we need a static converter? It shouldn't have any state. More generally, isn't there already some function in wxMac to do this conversion?

Reply:
> Sorry, I didn't really have time to look at this code yet, but why do we need a static converter? It shouldn't have any state. More generally, isn't there already some function in wxMac to do this conversion?

No hard reason, only the allocation, as we do so with wxMBConvUTF16 at other places in wx, I had just copied this. 

I have not found another place where I'm doing a UTF16 to wxString in osx code.

Reply:
> I'd suggest replacing this line with something along this, @vadz is this fine for you ?
> 
> ```
> static wxMBConvUTF16 s_UTF16Converter;
> itemString = wxString( static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId), s_UTF16Converter, dataSize );
> ```

Everything works!